crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,424 @@
+
+    agents[sender_idx]
+        .1
+        .broadcast(
+            space.clone(),
+            basis.clone(),
+            KitsuneTimeout::from_millis(5_000),
+            BroadcastData::Publish {
+                source: agents[sender_idx].2.clone(),
+                op_hash_list: vec![test_op.into()],
+                context: FetchContext::default(),
+            },
+        )
+        .await
+        .unwrap();
+
+    tokio::time::timeout(std::time::Duration::from_secs(60), {
+        let op_store_recv = agents[should_recv_idx].0.op_store().clone();
+        async move {
+            loop {
+                if !op_store_recv.read().is_empty() {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    assert_eq!(1, agents[should_recv_idx].0.op_store().read().len());
+
+    let mut agents_not_receiving_count = 0;
+    for i in 0..5 {
+        if i == sender_idx || i == should_recv_idx {
+            continue;
+        }
+
+        // We've filtered out the sender and the receiver, who are expected to have the data.
+        // Now we check that the agent at the current index does not have the basis that the op was
+        // published to in its arc. That would make the test wrong, not Kitsune, so fail here!
+        let should_this_agent_hold_the_op =
+            should_agent_hold_op_at_basis(&agents[i].0, agents[i].2.clone(), basis.clone());
+
+        if should_this_agent_hold_the_op {
+            tracing::warn!(
+                "Agent {} should not have received the data, this means that the test data construction isn't perfectly accurate but the test can continue",
+                i
+            );
+
+            continue;
+        }
+
+        // Now make the important assertion that the agent at index `i` did not receive the data! If it's not in the agents arc
+        // (which we just asserted above) then it should not have been received.
+        let store_lock = agents[i].0.op_store();
+        let store = store_lock.read();
+        assert!(
+            store.is_empty(),
+            "Agent {} should not have received any data but has {} ops. Ops store: {:?}",
+            i,
+            store.len(),
+            store,
+        );
+        agents_not_receiving_count += 1;
+    }
+
+    // A slightly looser assertion than requiring the test data to be constructed perfectly with no extra overlap.
+    assert!(
+        agents_not_receiving_count >= 2,
+        "At least two agents should not have received the data"
+    );
+}
+
+/// Very similar to the test above except the publisher is does not have the basis in its arc.
+/// This is a valid scenario because any hash might be produced by creating data and the publish
+/// should still go to the correct agents. It also says with the publisher, so we need to account
+/// for that when checking the op stores at the end of the test.
```

Comment:
```suggestion
/// Very similar to the test above except the publisher does not have the basis in its arc.
/// This is a valid scenario because any hash might be produced by creating data and the publish
/// should still go to the correct agents. It also stays with the publisher, so we need to account
/// for that when checking the op stores at the end of the test.
```

Reply:
Nice, thank you